docs/proposals/repo-server-multi-tenancy.md

```diff
@@ -0,0 +1,200 @@
+
+We propose to move this configuration into either the `argocd-cm` or to introduce a new Custom Resource, which would be consumed by the `argocd-server` and `argocd-application-controller`. The configuration for a single repository server instance would include the following items:
+
+```yaml=
+# A name for the instance. Will be used as back-reference. The name must
+# be compliant to RFC 1123 label names, as described in https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
+name: tenant-a
+# The DNS name or IP address of the repo-server's K8s service
+hostname: tenant-a-repo-server.argocd
+# The port of the repo-server service
+port: 8081
+# TLS configuration (optional)
+tls:
+  # Whether to ignore the remote TLS certificate 
+  insecure: false
+  # A root CA certificate to validate the TLS connection
+  caCert: |
+    ---- BEGIN CERTIFICATE ----
+    ...
+    ---- END CERTIFICATE ----
+
+```
+
+However, we do not propose to remove the appropriate command line options. The command line options (or their default values) will be used for the default repository server instance.
+
+### Mapping tenants to repository servers
+
+We propose to use the `AppProject` custom resource to map a project to a specific repository server instance. Given the above configuration example for a repository server instance named `tenant-a`, the `AppProject` configuration could simply look like the following:
+
+```yaml=
+kind: AppProject
+apiVersion: argoproj.io/v1alpha1
+metdata:
+  name: tenant-a-project-a
+spec:
+  repoServer: tenant-a
+```
+
+The new field `.spec.repoServer` will be an optional field in the `AppProject` spec.
+
+If the `.spec.repoServer` field is not specified in the `AppProject`, the default repository server (as specified implicit, or on the command line) will be used to generate manifests for Applications associated to this AppProject. This keeps the current behavior of Argo CD, and is fully backward compatible.
+
+If the `.spec.repoServer` field is specified, both `argocd-server` and `argocd-application-controller` would look up the appropriate configuration for the repository server instance to use, and create a gRPC clientset specifically for this instance when submitting requests to the repository server. When the referenced configuration is not found, there is no fallback to the default one. Instead, an error would be produced and manifest generation will fail. This is important for security reasons, so that a typo or misconfiguration would not lead to a potential flaw in isolation.
+
+The relationship of an `AppProject` to a repository server instance is `1:n`. An `AppProject` instance can use exactly one repository server instance, while each repository server instance can be referenced by multiple `AppProjects`. 
```

Comment:
hey @jannfis should this be `n:1` instead of `1:n`? The way I read it `1:n` suggests that one instance of AppProject could map to `n` instances of repository-server which is probably not going to be the case

Reply:
Good catch! You are correct, will fix that.